[Stable]Add helm chart for heptio/contour

jensoncs opened the pull request:
heptio/contour setup
enable autoscaling for contour

Co-authored-by: Jenson <jensonc8@gmail.com>

Co-authored-by: Sankalp <singh.sankalp1@gmail.com>

<!--
Thank you for contributing to helm/charts. Before you submit this PR we'd like to
make sure you are aware of our technical requirements and best practices:

* https://github.com/helm/charts/blob/master/CONTRIBUTING.md#technical-requirements
* https://github.com/helm/helm/tree/master/docs/chart_best_practices

For a quick overview across what we will look at reviewing your PR, please read
our review guidelines:

* https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md

Following our best practices right from the start will accelerate the review process and
help get your PR merged quicker.

When updates to your PR are requested, please add new commits and do not squash the
history. This will make it easier to identify new changes. The PR will be squashed
anyways when it is merged. Thanks.

For fast feedback, please @-mention maintainers that are listed in the Chart.yaml file.

Please make sure you test your changes before you push them. Once pushed, a CircleCI
will run across your changes and do some initial checks and linting. These checks run
very quickly. Please check the results. We would like these checks to pass before we
even continue reviewing your changes.
-->

**What this PR does / why we need it**:

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

jensoncs committed:
Add helm chart for heptio/contour

heptio/contour setup
enable autoscaling for contour

Co-authored-by: Jenson <jensonc8@gmail.com>

Co-authored-by: Sankalp <singh.sankalp1@gmail.com>
Signed-off-by: jenson <jensonc8@gmail.com>

davidkarlsen commented on stable/contour/templates/clusterrole.yaml:
```diff
@@ -0,0 +1,55 @@
+  labels:
+    app: {{ template "contour.fullname" . }}
+    chart: {{ template "contour.chart" . }}
+    release: {{ .Release.Name }}
```
add heritage as well, see: https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md

jensoncs replied:
Added heritage to metadata labels.

davidkarlsen commented on stable/contour/templates/crb.yaml:
```diff
@@ -0,0 +1,19 @@
+  name: contour
+subjects:
+  - kind: ServiceAccount
+    name: {{ template "contour.fullname" . }}
```
create a helper for the serviceaccountname, see: https://github.com/helm/helm/blob/master/docs/chart_best_practices/rbac.md and use that for consistency

jensoncs replied:
Created helper for service account.

davidkarlsen commented on stable/contour/values.yaml:
```diff
@@ -0,0 +1,60 @@
+  minReplicas: 2
+  maxReplicas: 15
+  targetCPUUtilizationPercentage: 70
+resources:
```
I can't find the reference for it now - but usually resources is kept commented, as it varies from deployment to deployment. Hints for values can be set but commented out with #

jensoncs replied:
```
minReplicas: 2
maxReplicas: 15
targetCPUUtilizationPercentage: 70
```
Commented the above resources. These resources are used in hpa.

jensoncs committed:
Add heritage to metadata label

Add serviceaccount helper
Comment resource values in values.yaml
Update Readme
Update chart version

Co-authored-by: Jenson <jensonc8@gmail.com>
Co-authored-by: Sankalp <singh.sankalp1@gmail.com>
Signed-off-by: jenson <jensonc8@gmail.com>

davidkarlsen commented on stable/contour/README.md:
```diff
@@ -0,0 +1,101 @@
+| `service.internalHttpsPort`           | Service internal https port                                  | 8443                                              |
+| `service.protocol`                    | Service Protocol                                             | TCP                                               |
+| `service.loadBalancerType`            | Service Load Balancer type                                   | LoadBalancer                                      |
+| `rbac.create`                         | Create rbac role for contour                                 | yes                                               |
```
yes|no vs true|false ?

davidkarlsen replied:
https://github.com/helm/helm/blob/master/docs/chart_best_practices/rbac.md
sorry for the late response - I forgot to actually submit the review in github

jensoncs reviewed (commented):
Updated chart with requested changes

Commit:
Change yes|no value to true|false
Co-authored-by: Jenson <jensonc8@gmail.com>
Co-authored-by: Sankalp <singh.sankalp1@gmail.com>

Signed-off-by: jenson <jensonc8@gmail.com>

jensoncs reviewed (commented):
Fix mentioned changes.

jensoncs reviewed (commented):
signoff commits

jensoncs committed:
Lock contour image version to 0.5.0
Co-authored-by: Jenson <jensonc8@gmail.com>
Co-authored-by: Sankalp <singh.sankalp1@gmail.com>

Signed-off-by: jenson <jensonc8@gmail.com>

jensoncs reviewed (commented):
Updated the changes

remmeier commented on stable/contour/templates/crb.yaml:
```diff
@@ -0,0 +1,20 @@
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: contour
```
here the role name should probably also include the release name as prefix

jensoncs replied:
Added the fix

remmeier commented on stable/contour/templates/service.yaml:
```diff
@@ -0,0 +1,24 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: {{ template "contour.service.fullname" . }}
```
had to switch to `name: {{ template "contour.fullname" . }}` to have a fully qualified name include release name

jensoncs replied:
Added the fix

remmeier reviewed (commented):
FYI: left two comments of things I had to change. In general looks great and would be nice to be merged into master given that envoy gets more and more traction.

jensoncs committed:
Merge remote-tracking branch 'charts/master'

jensoncs committed:
Add latest review changes related to service,clusterrole binding name

Signed-off-by: Jenson <jensonc8@gmail.com>

jensoncs reviewed (commented):
Add requested changes

mattfarina commented on stable/contour/README.md:
````diff
@@ -0,0 +1,101 @@
+To install the chart with the release name `my-release`:
+
+```bash
+$ helm install --name my-release .
````
This install command assumes it is a git checkout. Can you please update this to use the stable repository instead of `.`. That is for all places the commands are used.

jensoncs replied:
Fixed the Readme.

mattfarina commented on stable/contour/README.md:
```diff
@@ -0,0 +1,101 @@
+|---------------------------------------|-------------------------------------                         |---------------------------------------------------|
+| `deployment.replicas`                 | Number of replica sets                                       | 2                                                 |
+| `contour.image.repository`            | Repository for contour container image                       | gcr.io/heptio-images/contour                      |
+| `contour.image.tag`                   | Contour image tag                                            | master                                            |
```
The version of the image is in the values file as a tag. Can you please update this instead of noting `master`.

jensoncs replied:
Added contour version instead of master

mattfarina commented on stable/contour/templates/clusterrole.yaml:
```diff
@@ -0,0 +1,56 @@
+    app: {{ template "contour.fullname" . }}
+    chart: {{ template "contour.chart" . }}
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
```
I apologize for "moving the goal posts on you" but the format for the labels has changed because of the way upstream Kubernetes not recommends using them. See https://github.com/helm/charts/blob/master/REVIEW_GUIDELINES.md

jensoncs replied:
Fixed the labels according to k8s recommendations

mattfarina commented on stable/contour/templates/deployment.yaml:
```diff
@@ -0,0 +1,83 @@
+apiVersion: extensions/v1beta1
```
Please use the `apps/v1` API instead. the extensions one for Deployment will soon not be accepted by the Kubernetes API as it is long past the deprecation window.

jensoncs replied:
changed to apps/v1

mattfarina commented on stable/contour/values.yaml:
```diff
@@ -0,0 +1,63 @@
+  limits:
+    cpu: "400m"
+  requests:
+    cpu: "200m"
```
Can you please comment out the resource limits.

jensoncs replied:
commented the optional values

jensoncs committed:
Merge remote-tracking branch 'charts/master'

jensoncs committed:
Make lables in Kubernetes recommened way
- Fix Readme
Fix deployment api
comment optional hpa value

Signed-off-by: Jenson <jensonc8@gmail.com>

jensoncs reviewed (commented):
Added suggested changes

jensoncs committed:
Fix the linting issue

Signed-off-by: Jenson <jensonc8@gmail.com>

mattfarina commented on stable/contour/Chart.yaml:
```diff
@@ -0,0 +1,15 @@
+name: contour
+version: 0.1.6
+description: Contour envoy based ingress controller installation
+appVersion: "master"
```
The app version should be the version of the app rather than master. Is this the same value as the image tag?

jensoncs replied:
Changed app version to the image tag.

mattfarina commented on stable/contour/Chart.yaml:
```diff
@@ -0,0 +1,15 @@
+version: 0.1.6
+description: Contour envoy based ingress controller installation
+appVersion: "master"
+home: https://github.com/kubernetes/helm
```
Can you please link to Contour rather than Helm for the home

jensoncs replied:
Changed home URL to contour url

mattfarina commented on stable/contour/README.md:
```diff
@@ -0,0 +1,101 @@
+| `deployment.replicas`                 | Number of replica sets                                       | 2                                                 |
+| `contour.image.repository`            | Repository for contour container image                       | gcr.io/heptio-images/contour                      |
+| `contour.image.tag`                   | Contour image tag                                            | v0.5.0                                            |
+| `contour.image.pullPolicy`            | Image pull policy                                            | Always                                            |
```
Since a tag is being used can the `pullPolicy` be set to `IfNotPresent`. This way caching can be used in Kubernetes for the image.

jensoncs replied:
Changed image pullpolicy to IfNotPresent

mattfarina commented on stable/contour/templates/clusterrole.yaml:
```diff
@@ -0,0 +1,56 @@
+{{- if .Values.rbac.create -}}
+
+apiVersion: rbac.authorization.k8s.io/v1beta1
```
Instead of the beta version of the RBAC API can you please use the release version. The release version has been available long enough that even unsupported versions of Kubernetes support it. This applies to all the RBAC objects.

jensoncs replied:
Changed all RBAC API version to release version.

mattfarina commented on stable/contour/values.yaml:
```diff
@@ -0,0 +1,63 @@
+  image:
+    repository: gcr.io/heptio-images/contour
+    tag: v0.5.0
+    pullPolicy: Always
```
Same as the previous pullPolicy comment

jensoncs replied:
Changed image pullpolicy to IfNotPresent

mattfarina reviewed (commented):
Thank you for the chart submission. There are a handful of minor changes needed.

jensoncs committed:
Merge remote-tracking branch 'charts/master'

jensoncs committed:
Set contour URL in chart.yaml
- Update appversion
Update imagepull policy
Change RBAC api from beta to release version
Update chart version

Signed-off-by: Jenson <jensonc8@gmail.com>

jensoncs reviewed (commented):
Added requested changes.

mattfarina commented on stable/contour/values.yaml:
```diff
@@ -0,0 +1,63 @@
+contour:
+  image:
+    repository: gcr.io/heptio-images/contour
+    tag: v0.5.0
```
Can you please update to a newer version of contour. I see v0.9.0 has been released.

jensoncs replied:
@mattfarina Upgraded contour and envoy version to latest.

Commit:
Merge remote-tracking branch 'charts/master'

jensoncs committed:
Upgrade contour and envoy version to latest

Signed-off-by: Jenson <jensonc8@gmail.com>

Commit:
Merge remote-tracking branch 'charts/master'

jensoncs committed:
Fix clusterrole binding and service account name

Signed-off-by: Jenson <jensonc8@gmail.com>

Commit:
Merge remote-tracking branch 'charts/master'

unguiculus reviewed (changes requested):
Please make sure you have the complete set of standard labels on all resources.

mattfarina commented on stable/contour/Chart.yaml:
```diff
@@ -0,0 +1,15 @@
+name: contour
+version: 0.2.0
+description: Contour envoy based ingress controller installation
+appVersion: "v0.9.0"
```
v0.11.0 is out. Should this be updated? I do realize this is the result of the PR taking awhile.

mattfarina commented on stable/contour/templates/crd.yaml:
```diff
@@ -0,0 +1,157 @@
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
```
Should the CRDs use the [crd-install hook](https://helm.sh/docs/developing_charts/#hooks)?

mattfarina commented on stable/contour/values.yaml:
```diff
@@ -0,0 +1,63 @@
+contour:
+  image:
+    repository: gcr.io/heptio-images/contour
+    tag: v0.9.0
```
Should the newer v11 be used now?

mattfarina commented on stable/contour/values.yaml:
```diff
@@ -0,0 +1,63 @@
+envoy:
+  image:
+    repository: docker.io/envoyproxy/envoy-alpine
+    tag: v1.9.0
```
V1.10 is out now. Should that be used?
